### PR TITLE
Fix the `TemplateOptionsListener`

### LINKED ISSUE #6842: Module template selection wrong base directory
### Affected version(s)

5.3.0-RC2

### Description

Updating from Contao 5.2 to 5.3.0-RC2 the selected twig templates in modules of type "template" (Individuelles Template) from the folder frontend_module/template/ are declared "unknown" and "content_element/template" is suggested. This seems to be a bug.

Here an example:
example

### PR DESCRIPTION
**Pull request by fritzmg**
Fixes #6842 

This changes the `TemplateOptionsListener` to a single service instance, so that we can continue to use PHP attributes for tagging.

### CODE REVIEW

**Review by ausi — DISMISSED** [review]

**Review by m-vo — DISMISSED** [review]
♥️

File: core-bundle/src/DependencyInjection/Compiler/RegisterFragmentsPass.php [hunk 1]
```diff
@@ -131,8 +132,8 @@ protected function registerFragments(ContainerBuilder $container, string $tag):
         $this->addPreHandlers($container, $preHandlers);
         $this->addGlobalsMapListener($globals, $container);
 
-        if (null !== $this->templateOptionsListener && $container->hasDefinition($this->templateOptionsListener)) {
-            $container->findDefinition($this->templateOptionsListener)->addMethodCall('setDefaultIdentifiersByType', [$templates]);
+        if ($this->dca && null !== $this->templateOptionsListener && $container->hasDefinition($this->templateOptionsListener)) {
```
Comment (leofeyer) [inline core-bundle/src/DependencyInjection/Compiler/RegisterFragmentsPass.php]:
```suggestion
        if (null !== $this->dca && null !== $this->templateOptionsListener && $container->hasDefinition($this->templateOptionsListener)) {
```

**Comment by aschempp** [comment]
I'm unsure whether we should do this. The service is not marked `@internal` and could very likely be used for anyone to offer a dropdown. Same as people did with `Controller::getTemplateGroup`, I would register my own copy of this service if I need a dropdown of templates.

We should see this service as a library service that we are using in our bundle, where autoconfiguration would not make sense at all.

**Comment by fritzmg** [comment]
I agree with @aschempp that we cannot and should not fix it this way. This arbitrarily limits the listener to a certain set of DCAs and thus this service needs to live as several instances, meaning we cannot use PHP attributes here. Instead, we need to revert the changes from https://github.com/contao/contao/pull/6743 to the `TemplateOptionsListener`.

**Comment by leofeyer** [comment]
I disagree. Listeners are not covered by our BC promise as documented here: https://docs.contao.org/dev/internals/bc-promise/ And the solution has already been approved by three core developers.

**Comment by fritzmg** [comment]
> Listeners are not covered by our BC promise

Yes, but this listener is different, it should be seen as a library service as mentioned by @aschempp. This way you can create your own instance of this service for your own DCA rather than having to re-implement the whole listener yourself for your template selection fields. This PR would kill that ability.

**Comment by leofeyer** [comment]
I understand that, but we still have to fix the problem that the listener cannot be auto-registered. People might need to use it in their app and they will run into the same problem.

**Comment by fritzmg** [comment]
> People might need to use it in their app and they will run into the same problem.

But this PR would not allow that as the DCA tables would be hardcoded.

**Comment by leofeyer** [comment]
I know.

**Comment by m-vo** [comment]
IMHO you should **not** reuse this listener in your application. The abstraction to find the right templates is the `Twig\Finder` service - the other things we added to the listener are only BC layers.

**Comment by fritzmg** [comment]
> The abstraction to find the right templates is the `Twig\Finder` service - the other things we added to the listener are only BC layers.

The `Twig\Finder` service still relies on the identifier passed to the `TemplateOptionsListener` in our compiler pass though, does it not? And as such this listener isn't there strictly for BC.

**Review by ausi — DISMISSED** [review]

**Comment by leofeyer** [comment]
As discussed in the Contao call, we want to merge the PR, as the listener is very specific to the core bundle and not meant to be reusable.

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @fritzmg.

**Pull request merged**